internal/v4/api.go

```diff
@@ -1169,6 +1170,61 @@ func (h *ReqHandler) serveAdminPromulgate(id *router.ResolvedURL, w http.Respons
 	return nil
 }
 
+// PUT id/publish
+// See https://github.com/juju/charmstore/blob/v4/docs/API.md#put-idpublish
+func (h *ReqHandler) servePublish(id *charm.URL, w http.ResponseWriter, req *http.Request) error {
+	// Perform basic validation of the request.
+	if req.Method != "PUT" {
+		return errgo.WithCausef(nil, params.ErrMethodNotAllowed, "%s not allowed", req.Method)
+	}
+	if id.Channel == charm.DevelopmentChannel {
+		return errgo.WithCausef(nil, params.ErrForbidden, "cannot publish or unpublish development charm or bundle %q", id)
+	}
+	jsonContentType := "application/json"
+	if ct := req.Header.Get("Content-Type"); ct != jsonContentType {
+		return errgo.WithCausef(nil, params.ErrBadRequest, "unexpected Content-Type %q; expected %q", ct, jsonContentType)
+	}
+
+	// Retrieve the requested action from the request body.
+	var publish params.PublishRequest
+	decoder := json.NewDecoder(req.Body)
+	if err := decoder.Decode(&publish); err != nil {
```

Comment:
e.g.

    var publish struct {
        params.PublishRequest `httprequest:",body"`
    }
    err := httprequest.Unmarshal(httprequest.Params{Request: req}, &publish)

would probably do the job.